tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java

```diff
@@ -0,0 +1,402 @@
+package org.tensorflow.framework.losses.impl;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.losses.Reduction;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.AssertThat;
+import org.tensorflow.op.core.ReduceSum;
+import org.tensorflow.op.core.SetDiff1d;
+import org.tensorflow.op.core.Squeeze;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.family.TNumber;
+
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * These are helper methods for Losses and will be module private when Java modularity is applied to
+ * TensorFlow Java. These methods should not be used outside of the Loss package.
+ */
+public class LossesImpl {
```

Comment:
For me, a `*Impl` should be the implementation of an interface, this one looks more like a `LossesHelper` with all its static methods (and the class should probably be final).

I did not went through the whole thing but it looks like these helpers could also be moved directly to `Loss` as protected methods?

Reply:
The split really comes for module visibility. `Losses` should be publicly accessible, while `LossesImpl` should be module private. Some `LossesImpl` methods may be used by `metrics`. Whether we call it `LossesImpl` of `LossesHelper` is a matter of preference. The current methods in `LossesImpl` should not be restricted to Loss classes as `metrics` classes may also make use of them, therefore `protected` is not the right semantic, .

Reply:
It feels uncomfortable to me that we plan to use the `LossesImpl` methods from other parts of our framework while restricting them from public use. When a system's built-ins rely on privileged capabilities that aren't available to 3rd-party code, I think it is commonly a big problem for the system's extensibility. In this case, I do see room to argue that these methods aren't "capabilities", but are just "implementation" which can safely be hidden. But given that it is important to us to reuse them for our own metrics, I lean toward thinking of them as capabilities that we should expose.

Reply:
There is a tight symmetry between Losses and Metrics as many (but not all) metrics rely on the methods in Losses.
Don't think other packages will have this close of a relationship.

Reply:
Is there a potential use case justifying exposing these to the public? Seeing as they are utilities needed to implement Losses/Metrics. 

Agree with a rename to LossesHelper or LossesUtility to differentiate from interface implementation, however.